mountpoint-s3-fs/src/superblock.rs

```diff
@@ -321,6 +323,97 @@ impl<OC: ObjectClient + Send + Sync> Superblock<OC> {
         trace!("Added handle with id: {}", handle_id);
         Ok(handle_id)
     }
+
+    async fn start_reading(&self, looked_up_inode: LookedUpInode, fh: u64) -> Result<Lookup, InodeError> {
+        let completion_hook = {
+            let mut locked_inode = looked_up_inode.inode.get_mut_inode_state()?;
+            if matches!(locked_inode.write_status, WriteStatus::LocalUnopened)
+                || !self.inner.inode_handles.try_add_reader(&locked_inode, fh)
+            {
+                return Err(InodeError::InodeNotReadableWhileWriting(looked_up_inode.inode.err()));
+            }
+            if matches!(locked_inode.write_status, WriteStatus::LocalOpenForWriting) {
+                locked_inode.write_status = WriteStatus::Remote;
+            }
+            locked_inode.completion_hook.clone()
+        };
+
+        if let Some(hook) = completion_hook
+            && let Some(lookup) = hook.trigger().await?
+        {
+            Ok(lookup)
+        } else {
+            Ok(looked_up_inode.into())
+        }
+    }
+
+    async fn start_writing(
```

Comment:
Let's document what does this method allow. E.g. opening for writing if (and only if) all of the other file handles for this inode have been flushed.